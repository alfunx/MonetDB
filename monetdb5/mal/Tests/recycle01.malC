#test the basics of the recycler
#reusing select results

module sql;
function sql.mvc():int;
	return mvc:=1;
end mvc;

function sql.tid(mvc:int,s:str,t:str):bat[:oid];
b:= bat.new(:oid);
barrier i:= 0:oid;
    bat.append(b,i);
    redo i:= iterator.next(1:oid,12:oid);
exit i;
    return tid:= b;
end tid;

function sql.bind(mvc:int, s:str,t:str,c:str,i:int):bat[:int];
	b:= bat.new(:int);
	bat.append(b,0);
	bat.append(b,1);
	bat.append(b,2);
	bat.append(b,3);
	bat.append(b,4);
	bat.append(b,5);
	bat.append(b,6);
	bat.append(b,7);
	bat.append(b,8);
	bat.append(b,9);
	bat.append(b,10);
	bat.append(b,11);
	return bind:=b;
end bind;

function qry(low:int, hgh:int);
	m:= sql.mvc();
	x:bat[:oid]  := sql.tid(m,"sys","tbl");
	b:bat[:int]:= sql.bind(m,"sys","tbl","col",0);
	s:= algebra.subselect(b,x,low,hgh,true,true,false);
	io.print(s);
end qry;

optimizer.recycler("user","qry");
mdb.list("user","qry");
io.print("start");
qry(1,10);
io.print("reuse BAT");
qry(1,10);
recycle.dump();
