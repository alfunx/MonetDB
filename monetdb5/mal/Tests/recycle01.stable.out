stdout of test 'recycle01` in directory 'monetdb5/mal` itself:


# 20:44:21 >  
# 20:44:21 >  "mserver5" "--debug=10" "--set" "gdk_nr_threads=0" "--set" "mapi_open=true" "--set" "mapi_port=34866" "--set" "mapi_usock=/var/tmp/mtest-23471/.s.monetdb.34866" "--set" "monet_prompt=" "--forcemito" "--set" "mal_listing=2" "--dbpath=/export/scratch1/mk/default//Linux/var/MonetDB/mTests_monetdb5_mal" "--set" "mal_listing=2" "--set" "embedded_r=yes"
# 20:44:21 >  

# MonetDB 5 server v11.22.0
# This is an unreleased version
# Serving database 'mTests_monetdb5_mal', using 8 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs and 128bit integers dynamically linked
# Found 15.589 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2015 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://vienna.da.cwi.nl:34866/
# Listening for UNIX domain connection requests on mapi:monetdb:///var/tmp/mtest-23471/.s.monetdb.34866
# MonetDB/GIS module loaded
# Start processing logs sql/sql_logs version 52200
# Start reading the write-ahead log 'sql_logs/sql/log.4'
# Finished reading the write-ahead log 'sql_logs/sql/log.4'
# Finished processing logs sql/sql_logs
# MonetDB/SQL module loaded
# MonetDB/R   module loaded

Ready.
#RECYCLE drop
#RECYCLER CATALOG cached 4 instructions, MAL recycled = 4 total MAL executed = 83 memory= 144 total searchtime=30.75(usec) savings=3843736
# CL	   lru		cnt	 ticks	 rd	 wr	 Instr
#0	2	960904	0	0	#    1 := sql.mvc();
#1	2	960923	0	96	#    <tmp_34> := sql.tid(1,"sys","tbl");
#2	2	960943	0	48	#    <tmp_314> := sql.bind(1,"sys","tbl","col",0);
#3	2	960966	144	0	#    <tmp_436> := algebra.subselect(<tmp_314>,<tmp_34>,1,10,true,true,false);

# 20:44:21 >  
# 20:44:21 >  "mclient" "-lmal" "-ftest" "-Eutf-8" "--host=/var/tmp/mtest-23471" "--port=34866"
# 20:44:21 >  

function sql.qry(low:int,hgh:int):void;
    m := sql.mvc();
    x:bat[:oid] := sql.tid(m,"sys","tbl");
    b:bat[:int] := sql.bind(m,"sys","tbl","col",0);
    s := algebra.subselect(b,x,low,hgh,true,true,false);
    io.print(s);
end sql.qry;
[ "start"	]
#RECYCLE [  0] cost 5 mem 0 srch  -nan #    1 := sql.mvc();
#RECYCLE [  1] cost 64 mem 96 srch 27.00 #    <tmp_34> := sql.tid(1,"sys","tbl");
#RECYCLE [  2] cost 46 mem 48 srch 29.50 #    <tmp_314> := sql.bind(1,"sys","tbl","col",0);
#RECYCLE [  3] cost 22 mem 0 srch 30.00 #    <tmp_436> := algebra.subselect(<tmp_314>,<tmp_34>,1,10,true,true,false);
#--------------------------#
# h	t  # name
# void	void  # type
#--------------------------#
[ 0@0,	1@0	]
[ 1@0,	2@0	]
[ 2@0,	3@0	]
[ 3@0,	4@0	]
[ 4@0,	5@0	]
[ 5@0,	6@0	]
[ 6@0,	7@0	]
[ 7@0,	8@0	]
[ 8@0,	9@0	]
[ 9@0,	10@0	]
[ "reuse BAT"	]
#REUSED  [  0]  960904 (usec) #    1 := sql.mvc();
#REUSED  [  1]  960923 (usec) #    <tmp_34> := sql.tid(1,"sys","tbl");
#REUSED  [  2]  960943 (usec) #    <tmp_314> := sql.bind(1,"sys","tbl","col",0);
#REUSED  [  3]  960966 (usec) #    <tmp_436> := algebra.subselect(<tmp_314>,<tmp_34>,1,10,true,true,false);
#--------------------------#
# h	t  # name
# void	void  # type
#--------------------------#
[ 0@0,	1@0	]
[ 1@0,	2@0	]
[ 2@0,	3@0	]
[ 3@0,	4@0	]
[ 4@0,	5@0	]
[ 5@0,	6@0	]
[ 6@0,	7@0	]
[ 7@0,	8@0	]
[ 8@0,	9@0	]
[ 9@0,	10@0	]
#RECYCLER CATALOG cached 4 instructions, MAL recycled = 4 total MAL executed = 81 memory= 144 total searchtime=30.75(usec) savings=3843736
# CL	   lru		cnt	 ticks	 rd	 wr	 Instr
#0	2	960904	0	0	#    1 := sql.mvc();
#1	2	960923	0	96	#    <tmp_34> := sql.tid(1,"sys","tbl");
#2	2	960943	0	48	#    <tmp_314> := sql.bind(1,"sys","tbl","col",0);
#3	2	960966	144	0	#    <tmp_436> := algebra.subselect(<tmp_314>,<tmp_34>,1,10,true,true,false);

# 20:44:22 >  
# 20:44:22 >  "Done."
# 20:44:22 >  

